tools/reviewer/cas_proxy/Proxy.java

```diff
@@ -0,0 +1,200 @@
+
+      @Override
+      public void onNext(FileUploadRequest value) {
+        if (this.accessManager.hasWriteAccess(value.getIdToken())) {
+          hasAccess.set(true);
+        }
+
+        if (!hasAccess.get()) {
+          responseObserver.onError(new AccessDeniedException());
+        }
+
+        try {
+          buffer.write(value.getContent().toByteArray());
+        } catch (IOException e) {
+          e.printStackTrace();
+          responseObserver.onError(e);
+        }
+      }
+
+      @Override
+      public void onError(Throwable t) {
+        responseObserver.onError(t);
+      }
+
+      @Override
+      public void onCompleted() {
+        String sha256 = Hashing.sha256().hashBytes(buffer.toByteArray()).toString();
+        File tempFile;
+        try {
+          tempFile = File.createTempFile(sha256, ".tmp");
+          FileOutputStream outputStream = new FileOutputStream(tempFile);
+          outputStream.write(buffer.toByteArray());
+          outputStream.close();
+
+          protoClient.uploadTo(Proxy.CAS_BUCKET_NAME, tempFile.toPath().toString(), sha256);
+          responseObserver.onNext(FileUploadResponse.newBuilder().setSha256(sha256).build());
+        } catch (IOException e) {
+          e.printStackTrace();
+          responseObserver.onError(e);
+        }
+      }
+    }
+
+    @Override
+    public void downloadFile(
+        FileDownloadRequest request, StreamObserver<FileDownloadResponse> responseObserver) {
+
+      if (!this.accessManager.hasReadAccess(request.getIdToken(), request.getSha256())) {
+        responseObserver.onError(new AccessDeniedException());
+      }
+
+      String fileName =
+          fileCache.computeIfAbsent(
+              request.getSha256(),
+              (sha256) -> {
+                try {
+                  return client.downloadFrom(CAS_BUCKET_NAME, sha256);
+                } catch (IOException e) {
+                  responseObserver.onError(e);
+                  return null;
+                }
+              });
+
+      if (fileName == null) {
+        return;
+      }
+
+      try {
+        File file = new File(fileName);
+        FileInputStream stream = new FileInputStream(file);
+
+        byte[] chunk = new byte[1024];
+
+        int bytesRead;
+
+        while ((bytesRead = stream.read(chunk)) != -1) {
+          responseObserver.onNext(
+              FileDownloadResponse.newBuilder()
+                  .setContent(ByteString.copyFrom(chunk, 0, bytesRead))
+                  .build());
+        }
+
+        responseObserver.onCompleted();
+      } catch (IOException e) {
+        e.printStackTrace();
+        responseObserver.onError(e);
+      }
+    }
+
+    @Override
+    public StreamObserver<FileUploadRequest> uploadFile(
+        StreamObserver<FileUploadResponse> responseObserver) {
+      return new FileUploader(this.client, this.accessManager, responseObserver);
+    }
+
+    public Service(String serviceAccountPath, AccessManager accessManager) {
```

Comment:
Can you put the ctors on top?

Reply:
@oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c